src/cdk/tree/control/base-tree-control.ts

```diff
@@ -0,0 +1,85 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {SelectionModel} from '@angular/cdk/collections';
+import {Observable} from 'rxjs/Observable';
+import {TreeControl} from './tree-control';
+
+/** Base tree control. It has basic toggle/expand/collapse operations on a single data node. */
+export abstract class BaseTreeControl<T> implements TreeControl<T> {
+  /** Saved data node for `expandAll` action. */
+  dataNodes: T[];
+  /** A selection model with multi-selection to track expansion status. */
+  expansionModel: SelectionModel<T> = new SelectionModel<T>(true);
+
+  /**
+   * Toggles one single data node.
+   * Expands a collapsed data node or collapse an expanded data node.
+   */
+  toggle(dataNode: T): void {
+    this.expansionModel.toggle(dataNode);
+  }
+
+  /** Expands one single data node. */
+  expand(dataNode: T): void {
+    this.expansionModel.select(dataNode);
+  }
+
+  /** Collapses one single data node. */
+  collapse(dataNode: T): void {
+    this.expansionModel.deselect(dataNode);
+  }
+
+  /** Whether a given data node is expanded or not. Returns true if the data node is expanded. */
+  isExpanded(dataNode: T): boolean {
+    return this.expansionModel.isSelected(dataNode);
+  }
+
+  /** Toggles a subtree rooted at `node` recursively. */
+  toggleDescendants(dataNode: T): void {
+    this.expansionModel.isSelected(dataNode)
+        ? this.collapseDescendants(dataNode)
+        : this.expandDescendants(dataNode);
+  }
+
+  /** Collapse all dataNodes in the tree. */
+  collapseAll(): void {
+    this.expansionModel.clear();
+  }
+
+  /** Expands a subtree rooted at given data node recursively. */
+  expandDescendants(dataNode: T): void {
+    let toBeProcessed = [dataNode];
+    toBeProcessed.push(...this.getDescendants(dataNode));
+    this.expansionModel.select(...toBeProcessed);
+  }
+
+  /** Collapses a subtree rooted at given data node recursively. */
+  collapseDescendants(dataNode: T): void {
+    let toBeProcessed = [dataNode];
+    toBeProcessed.push(...this.getDescendants(dataNode));
+    this.expansionModel.deselect(...toBeProcessed);
+  }
+
+  /** Gets a list of descendent data nodes of a subtree rooted at given data node recursively. */
```

Comment:
This is subjective, but it'd make sense to me for abstract methods to be nearer to the top to make it clear what needs to be defined when I extend this class.